src/waggle/data/vision.py

```diff
@@ -135,34 +194,87 @@ def __enter__(self):
                 raise RuntimeError(
                     f"unable to open video capture for device {self.device!r}"
                 )
+            # spin up a thread to keep up with the camera frame rate
+            if self.enable_daemon:
+                self.daemon.start()
         self.context_depth += 1
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.context_depth -= 1
         if self.context_depth == 0:
+            if self.enable_daemon:
+                self.daemon_need_to_stop = True
+                self.daemon.join()
             self.capture.release()
-
-    def snapshot(self):
-        ok = self.capture.grab()
-        if not ok:
-            raise RuntimeError("failed to take snapshot")
-        timestamp = get_timestamp()
-        ok, data = self.capture.retrieve()
-        if not ok:
-            raise RuntimeError("failed to retrieve the taken snapshot")
-        return ImageSample(data=data, timestamp=timestamp, format=self.format)
-
-    def stream(self):
-        while True:
+    
+    def _run(self):
+        # we sleep slighly shorter than FPS to drain the buffer efficiently
+        fps = self.capture.get(cv2.CAP_PROP_FPS)
+        sleep = 0.01
+        if fps > 0 and fps < 100:
+            sleep = 1 / (fps + 1)
+        logging.debug(f'camera FPS is {fps}. the background thread sleeps {sleep} seconds in between grab()')
+        while not self.daemon_need_to_stop:
+            self.lock.acquire()
             ok = self.capture.grab()
             if not ok:
-                break
+                pass
+                # TODO: what should we do when grab() fails?
+                # raise RuntimeError("failed to grab a frame")
+            self.timestamp = get_timestamp()
+            self.lock.release()
+            if self._ready_for_next_frame.empty():
```

Comment:
I would use a:
```python
try:
    self._ready_for_next_frame.put_nowait()
except queue.Full
   pass
```

instead. You should never run into the issue of this changing between the check and put, since you're producer is also the only thing putting items here but you can basically do the put or ignore if already full instead.

You could also use a threading.Event here, if you wanted and just turn this into `event.set()` and have the consumer clear it.

Reply:
I updated the code block with `threading.Event`